sdk/nodejs/resource.ts

```diff
@@ -384,26 +389,58 @@ export class Output<T> {
-                // give us an actual value for this Output.
-                const perform = await isKnown;
-                if (runtime.isDryRun() && !perform) {
-                    return <U><any>undefined;
+                try {
+                    if (runtime.isDryRun()) {
+                        // During previews only perform the apply if the engine was able to
+                        // give us an actual value for this Output.
+                        const applyDuringPreview = await isKnown;
+
+                        if (!applyDuringPreview) {
+                            // We didn't actually run the function, our new Output is definitely
+                            // **not** known.
+                            innerIsKnownResolve(false);
+                            return <U><any>undefined;
+                        }
+                    }
+
+                    const transformed = await func(v);
+                    if (Output.isInstance(transformed)) {
+                        // Note: if the func returned a Output, we unwrap that to get the inner value
+                        // returned by that Output.  Note that we are *not* capturing the Resources of
+                        // this inner Output.  That's intentional.  As the Output returned is only
+                        // supposed to be related this *this* Output object, those resources should
+                        // already be in our transitively reachable resource graph.
+
+                        // The callback func has produced an inner Output that may be 'known' or 'unknown'.
+                        // We have to properly forward that along to our outer output.  That way the Outer
+                        // output doesn't consider itself 'known' then the inner Output did not.
+                        innerIsKnownResolve(await transformed.isKnown);
+                        return await transformed.promise();
+                    } else {
+                        // We successfully ran the inner function.  Our new Output should be considered known.
+                        innerIsKnownResolve(true);
+                        return transformed;
+                    }
                 }
-
-                const transformed = await func(v);
-                if (Output.isInstance(transformed)) {
-                    // Note: if the func returned a Output, we unwrap that to get the inner value
-                    // returned by that Output.  Note that we are *not* capturing the Resources of
-                    // this inner Output.  That's intentional.  As the Output returned is only
-                    // supposed to be related this *this* Output object, those resources should
-                    // already be in our transitively reachable resource graph.
-                    return await transformed.promise();
-                } else {
-                    return transformed;
+                finally {
```

Comment:
Isn’t this a catch?  We don’t want to resolve the promise twice do we? (Not sure what the semantics of that are - maybe it’s okay and actually intentional here?)

Reply:
it was intentional.  Promise resolution works well, with all the resolve/reject methods working as "first caller wins".  So it's safe to do this.  I'll even add a test where 'isknown' is a rejected promise to demonstrate this.

This finally helps ensure that no matter what we don't leave that promise unhandled.